### ISSUE TITLE
A XSS is not detect by ActiveScan - Other tools use ML

### ISSUE THREAD

**Issue by Prakhash** [zaproxy/zaproxy#3247]
Hi all,

I have an application which has XSS issues in 2 places which I manually tested and confirmed, both are reflected XSS issues and ZAP didn't detect the issues when I did the scan.

During both cases, I have even done the XSS while proxying through ZAP and the script was appeared on the source page as below but ZAP didn't recognize it(attack happened).

![screenshot from 2017-02-28 19-01-02](https://cloud.githubusercontent.com/assets/4003151/23407408/cfb6add8-fde9-11e6-95f1-6995e4cb1171.png)

_In this case, the attack is not straight forward as I have added the input validation fields. So after filling the input boxes with correct values I have intercepted the request and modified the values with scripts and it was executed in the browser._

During the 2nd case, I got the below exception in browser in browser and again ZAP failed to recognize the attack (Attack happened when I turned off proxying)

![screenshot from 2017-02-28 19-02-49](https://cloud.githubusercontent.com/assets/4003151/23407507/432e627e-fdea-11e6-966c-a0628ac0f49a.png)

In this case, the attack is very straight forward. The URL is something similar to below.
```
 https://localhost/test.jsp?name=DefaultApplication">Go Back</a><script>alert(1990)</script><"
```

What might be the reasons for this?

Also please specify the location in ZAP project where we can find this rules set to identify the XSS

Thanks

<sub>Edited 20170228 (by @kingthorin): Formatting tweaks, for readability.</sub>

**Comment by kingthorin** [comment]
> I have an application which has XSS issues in 2 places which I manually tested and confirmed, both are reflected XSS issues and ZAP didn't detect the issues when I did the scan.

It's probably just not an injection that value or output condition that is checked. Did you have the XSS rules set to insane?

> During both cases, I have even done the XSS while proxying through ZAP and the script was appeared on the source page as below but ZAP didn't recognize it(attack happened).

You expected ZAP to passively recognize your attack? That's an interesting idea but I can't see a way for that to happen, a human could inject anything they wanted to any effect they can foresee and there's no way for ZAP to be aware of any of those conditions.

> During the 2nd case, I got the below exception in browser in browser and again ZAP failed to recognize the attack (Attack happened when I turned off proxying)

You can prevent the use of chunked encoding by setting HTTP/1.0. More info here: https://groups.google.com/forum/#!searchin/zaproxy-develop/chunked%7Csort:relevance/zaproxy-develop/xBfXao472Ms/q3H0gdyHAgAJ

**Comment by Prakhash** [comment]
> It's probably just not an injection that value or output condition that is checked. Did you have the XSS rules set to insane?
Yes, I have done this.

> You expected ZAP to passively recognize your attack? That's an interesting idea but I can't see a way for that to happen, a human could inject anything they wanted to any effect they can foresee and there's no way for ZAP to be aware of any of those conditions.
My expectation wasn't the ZAP to passively recognize the attack(It would be also nice if we would have a function like this), I ran the active scan after this. AFAIK we should be having manual navigation up to some extend and ZAP will learn & build the URL tree from this, so I thought it will add our input parameters to its list and using it during the active scan too.(Correct me if I'm wrong)

> You can prevent the use of chunked encoding by setting HTTP/1.0. More info here: https://groups.google.com/forum/#!searchin/zaproxy-develop/chunked%7Csort:relevance/zaproxy-develop/xBfXao472Ms/q3H0gdyHAgAJ

Why is it only failing to the URL with the attack param? other URLs are working perfectly.

Also, these two attacks are initially identified by the tool Qualys, but I usually use ZAP for the tests, so I wanted to check where the ZAP is actually failing to identify this. 

It would be great if you could help me to figure out this issue and do the fixes(Maybe in ZAP) 

Thanks

**Comment by kingthorin** [comment]
@prakhash Your injection might be breaking protocol compliance which in turn is making zap interpret that chunking is in use even when it isn't really. There's not much we can do to address that.

As for XSS example 1 in your original ticket, which scanning rules do you have installed? Only release?

**Comment by kingthorin** [comment]
@prakhash any further details?

**Comment by Prakhash** [comment]
@kingthorin  Sorry for the late response. 

Yes I was using the released version. But now I tried the same scenario with the latest ZAP 2.6 release and it worked perfectly(Identified the issue), But still I was able to figure out some issues which were identified by some other tools and not by ZAP(2.6).

But when further analyzed I have figured out these tools use some ML mechanism to identify issues if we did it manually too.

For example,as per my understanding, if one place is not get detected by tools, if we say that manually as this is vulnerable, tool is capable of identifying similar scenarios to that.(In addition to the rules which we can improve)

I wish ZAP also should integrate functionalities like that.
